File: wk3/customSortString.java
// problem link: https://leetcode.com/problems/custom-sort-string/description/

class Solution {
    public String customSortString(String order, String s) {
        HashMap<Character, Integer> map = new HashMap<>();
        int ind = 0;
        for(char c : order.toCharArray())
            map.put(c, ind++);
        for(char c : s.toCharArray())
            if(!map.containsKey(c))
                map.put(c, -1);
        PriorityQueue<Character> pq = new PriorityQueue<>((a, b) -> map.get(a) - map.get(b));
        for(char c : s.toCharArray())
            pq.add(c);
        String ret = "";
        while(!pq.isEmpty())
            ret += pq.poll();
        return ret;
    }
}

File: wk3/cpp/customSortString.cpp
// problem link: https://leetcode.com/problems/custom-sort-string/description/

#include <bits/stdc++.h>
using namespace std;

class Solution {
public:
    string customSortString(string order, string s) {
        int arr[26] = {0};
        for(char& c : s)
            arr[c - 'a']++;
        string ret;
        for(char& c : order) {
            ret.append(string(arr[c - 'a'], c));
            arr[c - 'a'] = 0;
        }
        for(int i = 0; i < 26; i++)
            ret.append(string(arr[i], 'a' + i));
        return ret;
    }
};